src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -644,12 +659,14 @@ public void setEnclosingType(Type outer) {
                             sbp - startSbp);
                     if (allowPrimitiveClasses && name.toString().endsWith("$ref")) {
                         name = name.subName(0, name.length() - 4);
-                        requireProjection = true;
+                        Assert.check(prefix == 'L');
+                        flavor = L_TypeOf_Q;
                     } else {
-                        requireProjection = false;
+                        // We are seeing QFoo; or LFoo; The name itself does not shine any light on default val-refness
+                        flavor = prefix == 'L' ? Flavor.L_TypeOf_X : Flavor.L_TypeOf_X;
```

Comment:
Is there a typo here - shouldn't be:

```
                        flavor = prefix == 'L' ? Flavor.L_TypeOf_X : Flavor.Q_TypeOf_X;
```

Reply:
Oops. Yes it should be as you say. I am puzzled because this was copy pasted code. Don't know how the error got introduced! Good catch!